### ISSUE TITLE
Add a metadata envelope attribute for GitHub repo

### ISSUE THREAD

**Issue by smashwilson** [deconst/deconst-docs#158]
We want to support "edit this on GitHub" links in the final content. To support this, we'll need to add an (optional) attribute to the metadata envelope spec to store which repository a particular page came from.

**Comment by ktbartholomew** [comment]
I think perhaps we should consider generalizing this a bit to cover the need to define/present repository-wide information. This includes the edit on GitHub link, but could also include the title of a repository, a list of contributors, taxonomies, etc.

In fact, we already (sometimes) have this data point in the form of the `ContentIDBase` attribute in `_deconst.json`. We could build on this by defining a minimal schema that `_deconst.json` should follow, as well as leaving room for a "meta" dictionary that would allow us or other implementers of deconst to extend it in an open-ended way.

I could also see a use case in which a preparer or build step modifies `_deconst.json` to insert information that has to be programmatically derived from the content itself, or from an external source.

**Comment by meker12** [comment]
Will we need to add anything to the RST source to enable the Github link?

**Comment by smashwilson** [comment]
> Will we need to add anything to the RST source to enable the Github link?

Nah, we should be able to take care of this in the site templates in the control repository.

> I think perhaps we should consider generalizing this a bit to cover the need to define/present repository-wide information. This includes the edit on GitHub link, but could also include the title of a repository, a list of contributors, taxonomies, etc.

Yeah, definitely. I think it's important to keep the metadata envelope spec (a) explicit (documented and validated, as opposed to ad-hoc) and (b) consistent (from preparer to preparer). Within those constraints there's a lot of room for growth, and these would be great things to include.

#133 would be a good candidate for doing this efficiently: if we're accepting a tarball, we can streamline the process for associating content repository-wide metadata like this as well as the TOC.

> In fact, we already (sometimes) have this data point in the form of the ContentIDBase attribute in _deconst.json. We could build on this by defining a minimal schema that _deconst.json should follow, as well as leaving room for a "meta" dictionary that would allow us or other implementers of deconst to extend it in an open-ended way.

I've tried hard to avoid overloading the meaning of the content ID as much as possible because, while the existing ones happen to be related to the GitHub repository URL, we already have content we want to include that isn't hosted on GitHub. We _could_ do shenanigans to recognize when the `contentIDBase` "looks like" a GitHub repository, but I prefer to keep things explicit whenever possible. We already get enough confusion about content IDs :wink:

:+1: on a schema for `_deconst.json` files. We technically already have one (the preparers warn when they see invalid keys and fail when missing the required one) but it should be documented just like the metadata envelope schema. Also :+1: on using that as the source for repository-wide metadata, although maintaining things like contributor lists there would be cumbersome. Maybe the preparers could key off of a GitHub repo address, if one is provided, and use the GitHub API to gather information automatically?

I do like carving out a place for "nonstandard" keys in the spec, both here and for metadata envelopes, although I'm not fond of "meta" as a name for it. A little too generic for my tastes, I think. "ext" maybe?

> I could also see a use case in which a preparer or build step modifies _deconst.json to insert information that has to be programmatically derived from the content itself, or from an external source.

:-1: Let's not go down the road of self-modifying configuration files. I say put just enough metadata in `_deconst.json` to derive everything, then keep the rest of the work in the preparers at build time. If we do need to cache that I feel that should be part of the build process.

**Comment by meker12** [comment]
@smashwilson  and @ktbartholomew Sphinx has an extension that adds the Edit on github link to pages -- Can you leverage to provide this functionality:  https://pypi.python.org/pypi/sphinxcontrib-srclinks

If we added this capability to our Sphinx projects, would it be supported when docs were built using the Sphinx preparer?

**Comment by meker12** [comment]
The edit on github feature has been on our feature list for quite awhile now.  Can you provide an idea on how long it would take to implement this feature.   @smashwilson 

Added Trello card with info about requirements:  https://trello.com/c/ExyBSA8e

**Comment by smashwilson** [comment]
The answer is that @ktbartholomew did the backend work a month ago in:

deconst/preparer-jekyll#30
deconst/preparer-jekyll#29
deconst/preparer-sphinx#44
deconst/preparer-sphinx#45
deconst/preparer-sphinx#46

Last I heard we were waiting on input from design to add these to the templates as part of a larger design overhaul, but he could tell you better than me.

**Comment by ktbartholomew** [comment]
Yep, if this is a hot item then Christie Quaranto and team need to be aware so they can allocate some time to work on the UX side of things.

**Comment by meker12** [comment]
@ktbartholomew  So the work for DevDocs is the Sphinx preparer.  Are you able to leverage the Jekyll work for the Sphinx preparer, or is that a separate effort?

Also, I think the original wireframes had the Edit on github feature in them, so maybe that can be resurrected and implemented, until we have time for the larger overhaul.  Will follow up with Laura about that.

**Comment by ktbartholomew** [comment]
> So the work for DevDocs is the Sphinx preparer. Are you able to leverage the Jekyll work for the Sphinx preparer, or is that a separate effort?

I'm not sure what you're asking. both Jekyll and Sphinx repos are capable of producing "Edit on Github" metadata: https://deconst.horse/writing-docs/author/#content-repository-metadata

**Comment by meker12** [comment]
In an earlier comment, the links were going to the Jekyll repo...

deconst/preparer-jekyll#30
deconst/preparer-jekyll#29
deconst/preparer-sphinx#44
deconst/preparer-sphinx#45
deconst/preparer-sphinx#46

So wanted to confirm that work for Sphinx is there, too.  Thanks for the confirmation.

**Comment by meker12** [comment]
## Current implementation

```
githubUrl: The content repository’s Github URL. If present, will be used to add 
github_issues_url and github_edit_url to each submitted content envelope.
```

_Edit on github_ takes user to the top-level of the repo with the doc source.  User has to navigate to find file to edit. 

For some projects, where the doc source might be deeply nested, users might have difficulty finding the right file,  for example see the [Autoscale repo](https://github.com/rackerlabs/otter).

## Desired implementation

_Edit on github_ link  takes users to the Sphinx project directory where they can find the file to edit. 

**Requirements**
- Enable metadata that provides the path to the Sphinx project root directory.  
- In templates, use doc root value as the target for the _Edit on github_ link. 

For example, `githubDocRoot: https://github.com/rackerlabs/docs-cloud-images/tree/master/api-docs`

## Cadillac implementation that makes it super-convenient for someone to contribute

**Edit on github** takes user directly to the file they want to edit.  

### Requirement
- Append the source file name to the githubDocRoot URL.  
- In templates, use the concatenated value as the target for the **Edit on github** link. 

Local Sphinx json build provides the source file name (with *.txt instead of *.rst). Not sure what happens with Sphinx preparer build.

```
{
    "body": "<div class=\"section\" id=\"contract-versions\">\n<span id=\"id1\"></span><h1>Contract versions<a class=\"headerlink\" href=\"#contract-versions\" title=\"Permalink to this headline\">#</a></h1>\n<p>The contract version denotes the data model and behavior that the API\nsupports. The Identity API uses a URI version scheme.</p>\n<div class=\"admonition important\">\n<p class=\"first admonition-title\">Important</p>\n<p>The Rackspace implementation of the OpenStack Keystone Identity Service\nv2.0 does not yet support MIME-type version.</p>\n<p class=\"last\">All client requests to the API must specify the target version\nidentifier in the first element of the URI path. For example, v2.0 is\nthe target version in this enddpoint URI:</p>\n</div>\n<div class=\"code highlight-python\"><div class=\"highlight\"><pre>https://identity.api.rackspacecloud.com/v2.0/\n</pre></div>\n</div>\n<p>The only access available without the target version are to check for\nthe GET version API request that returns available API versions\nsupported at the specified endpoint.</p>\n<p>The folloiwng example shows an XML request that uses URI versioning.</p>\n<p><strong>Example: Request with URI version scheme</strong></p>\n<div class=\"code highlight-python\"><div class=\"highlight\"><pre>GET /v2.0/tenants HTTP/1.1\nHost: identity.api.rackspacecloud.com\nAccept: application/xml\nX-Auth-Token: eaaafd18-0fed-4b3a-81b4-663c99ec1cbb\n</pre></div>\n</div>\n<p>The URI scheme for API contracts supports the following development\npractices:</p>\n<ul class=\"simple\">\n<li>New features and functionality that do not break API compatibility\nare introduced in the current version of the API as\n<a class=\"reference internal\" href=\"../extensions/#extensions-ovw\"><span>extensions</span></a>. The URI version stays the\nsame. For example, the Identity service is developing the RAX-AUTH\nextension to develop new services and resources to support the\nRackspace implementation of OpenStack Keystone v2.0.</li>\n<li>Service providers must change the API contract version anytime they\nintroduce or deprecate features or functionality that make the API\nincompatible with previously released versions. When the version\nchanges, the target version in the URI is updated, from v2.0 to v3.0\nfor example.</li>\n<li>When new API versions are released, older versions are\n<code class=\"docutils literal\"><span class=\"pre\">DEPRECATED</span></code>. Because a service endpoint can host multiple API\nversions, service providers can work with developers and partners to\nensure adequate time to migrate to the new version before removing\nthe deprecated version from service.</li>\n</ul>\n<p>You can programmatically determine available API versions by performing\na <strong>GET</strong> on the root URL returned from the authentication system. In\nthe root URL, the version and everything to the right of it is\ntruncated. An Atom representation of the version&#8217;s resources is\nsupported when issuing a request with the <code class=\"docutils literal\"><span class=\"pre\">Accept</span></code> header containing\n<code class=\"docutils literal\"><span class=\"pre\">application/atom+xml</span></code> or by adding <code class=\"docutils literal\"><span class=\"pre\">.atom</span></code> to the request URI. This\nallows standard Atom clients to track version changes.</p>\n<p><strong>Example&nbsp;3.30.&nbsp;Versions List Request</strong></p>\n<div class=\"code highlight-python\"><div class=\"highlight\"><pre>GET HTTP/1.1\nHost: identity.api.rackspacecloud.com\n</pre></div>\n</div>\n<p>Normal Response Code(s): 200, 203</p>\n<p>Error Response Code(s): badRequest (400), identityFault (500),\nserviceUnavailable(503)</p>\n<p>This operation does not require a request body.</p>\n<p><strong>Example&nbsp;3.31.&nbsp;Versions List Response: XML</strong></p>\n<div class=\"code highlight-python\"><div class=\"highlight\"><pre>&lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot; standalone=&quot;yes&quot;?&gt;\n\n&lt;versions xmlns=&quot;http://docs.openstack.org/common/api/v1.0&quot; xmlns:atom=&quot;http://www.w3.org/2005/Atom&quot; xmlns:ns3=&quot;http://docs.rackspace.com/core/event&quot;&gt;\n\n        &lt;version updated=&quot;2011-07-19T22:30:00Z&quot; status=&quot;DEPRECATED&quot; id=&quot;v1.0&quot;&gt;\n                &lt;atom:link href=&quot;https://identity.api.rackspacecloud.com/v1.0&quot; rel=&quot;self&quot;/&gt;\n        &lt;/version&gt;\n\n        &lt;version updated=&quot;2012-01-19T22:30:00.25Z&quot; status=&quot;CURRENT&quot; id=&quot;v1.1&quot;&gt;\n                &lt;atom:link href=&quot;https://identity.api.rackspacecloud.com/v1.1/&quot; rel=&quot;self&quot;/&gt;\n                &lt;atom:link href=&quot;http://docs.rackspacecloud.com/auth/api/v1.1/auth-client-devguide-latest.pdf&quot; rel=&quot;describedby&quot; type=&quot;application/pdf&quot;/&gt;\n                &lt;atom:link href=&quot;http://docs.rackspacecloud.com/auth/api/v1.1/auth.wadl&quot; rel=&quot;describedby&quot; type=&quot;application/vnd.sun.wadl+xml&quot;/&gt;\n        &lt;/version&gt;\n\n        &lt;version updated=&quot;2012-01-19T22:30:00.25Z&quot; status=&quot;CURRENT&quot; id=&quot;v2.0&quot;&gt;\n                &lt;atom:link href=&quot;https://identity.api.rackspacecloud.com/v2.0/&quot; rel=&quot;self&quot;/&gt;\n                &lt;atom:link href=&quot;http://docs.rackspacecloud.com/auth/api/v2.0/auth-client-devguide-latest.pdf&quot; rel=&quot;describedby&quot; type=&quot;application/pdf&quot;/&gt;\n                &lt;atom:link href=&quot;http://docs.rackspacecloud.com/auth/api/v2.0/auth.wadl&quot; rel=&quot;describedby&quot; type=&quot;application/vnd.sun.wadl+xml&quot;/&gt;\n        &lt;/version&gt;\n\n&lt;/versions&gt;\n</pre></div>\n</div>\n<p><strong>Example&nbsp;3.32.&nbsp;Versions List Response: Atom</strong></p>\n<div class=\"code highlight-python\"><div class=\"highlight\"><pre>&lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot;?&gt;\n&lt;feed xmlns=&quot;http://www.w3.org/2005/Atom&quot;&gt;\n    &lt;title type=&quot;text&quot;&gt;Available API Versions&lt;/title&gt;\n    &lt;updated&gt;2010-12-12T18:30:02.25Z&lt;/updated&gt;\n   &lt;id&gt;http://identity.api.rackspacecloud.com/&lt;/id&gt;\n    &lt;author&gt;&lt;name&gt;Rackspace&lt;/name&gt;&lt;uri&gt;http://www.rackspace.com/&lt;/uri&gt;&lt;/author&gt;\n   &lt;link rel=&quot;self&quot; href=&quot;http://identity.api.rackspacecloud.com/&quot;/&gt;\n    &lt;entry&gt;\n       &lt;id&gt;http://identity.api.rackspacecloud.com/v2.0/&lt;/id&gt;\n       &lt;title type=&quot;text&quot;&gt;Version v2.0&lt;/title&gt;\n       &lt;updated&gt;2011-05-27T20:22:02.25Z&lt;/updated&gt;\n       &lt;link rel=&quot;self&quot; href=&quot;http://identity.api.rackspacecloud.com/v2.0/&quot;/&gt;\n       &lt;content type=&quot;text&quot;&gt;Version v2.1 CURRENT (2011-05-27T20:22:02.25Z)&lt;/content&gt;\n    &lt;/entry&gt;\n    &lt;entry&gt;\n       &lt;id&gt;http://identity.api.rackspacecloud.com/v1.1/&lt;/id&gt;\n       &lt;title type=&quot;text&quot;&gt;Version v1.1&lt;/title&gt;\n       &lt;updated&gt;2010-12-12T18:30:02.25Z&lt;/updated&gt;\n       &lt;link rel=&quot;self&quot; href=&quot;http://identity.api.rackspacecloud.com/v1.1/&quot;/&gt;\n       &lt;content type=&quot;text&quot;&gt;Version v1.1 CURRENT (2010-12-12T18:30:02.25Z)&lt;/content&gt;\n    &lt;/entry&gt;\n    &lt;entry&gt;\n       &lt;id&gt;http://identity.api.rackspacecloud.com/v1.0/&lt;/id&gt;\n       &lt;title type=&quot;text&quot;&gt;Version v1.0&lt;/title&gt;\n       &lt;updated&gt;2009-10-09T11:30:00Z&lt;/updated&gt;\n       &lt;link rel=&quot;self&quot; href=&quot;http://identity.api.rackspacecloud.com/v1.0/&quot;/&gt;\n       &lt;content type=&quot;text&quot;&gt;Version v1.0 DEPRECATED (2009-10-09T11:30:00Z)&lt;/content&gt;\n    &lt;/entry&gt;\n&lt;/feed&gt;\n</pre></div>\n</div>\n<p><strong>Example&nbsp;3.33.&nbsp;Versions List Response: JSON</strong></p>\n<div class=\"code highlight-python\"><div class=\"highlight\"><pre><span class=\"p\">{</span>\n    <span class=\"s\">&quot;versions&quot;</span><span class=\"p\">:</span> <span class=\"p\">{</span>\n        <span class=\"s\">&quot;version&quot;</span><span class=\"p\">:</span> <span class=\"p\">[</span>\n            <span class=\"p\">{</span>\n                <span class=\"s\">&quot;id&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;v1.0&quot;</span><span class=\"p\">,</span>\n                <span class=\"s\">&quot;link&quot;</span><span class=\"p\">:</span> <span class=\"p\">{</span>\n                    <span class=\"s\">&quot;href&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;https://identity.api.rackspacecloud.com/v1.0&quot;</span><span class=\"p\">,</span>\n                    <span class=\"s\">&quot;rel&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;self&quot;</span>\n                <span class=\"p\">},</span>\n                <span class=\"s\">&quot;status&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;DEPRECATED&quot;</span><span class=\"p\">,</span>\n                <span class=\"s\">&quot;updated&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;2011-07-19T22:30:00Z&quot;</span>\n            <span class=\"p\">},</span>\n            <span class=\"p\">{</span>\n                <span class=\"s\">&quot;id&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;v1.1&quot;</span><span class=\"p\">,</span>\n                <span class=\"s\">&quot;link&quot;</span><span class=\"p\">:</span> <span class=\"p\">{</span>\n                    <span class=\"s\">&quot;href&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;http://docs.rackspacecloud.com/auth/api/v1.1/auth.wadl&quot;</span><span class=\"p\">,</span>\n                    <span class=\"s\">&quot;rel&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;describedby&quot;</span><span class=\"p\">,</span>\n                    <span class=\"s\">&quot;type&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;application/vnd.sun.wadl+xml&quot;</span>\n                <span class=\"p\">},</span>\n                <span class=\"s\">&quot;status&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;CURRENT&quot;</span><span class=\"p\">,</span>\n                <span class=\"s\">&quot;updated&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;2012-01-19T22:30:00.25Z&quot;</span>\n            <span class=\"p\">},</span>\n            <span class=\"p\">{</span>\n                <span class=\"s\">&quot;id&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;v2.0&quot;</span><span class=\"p\">,</span>\n                <span class=\"s\">&quot;link&quot;</span><span class=\"p\">:</span> <span class=\"p\">{</span>\n                    <span class=\"s\">&quot;href&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;http://docs.rackspacecloud.com/auth/api/v2.0/auth.wadl&quot;</span><span class=\"p\">,</span>\n                    <span class=\"s\">&quot;rel&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;describedby&quot;</span><span class=\"p\">,</span>\n                    <span class=\"s\">&quot;type&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;application/vnd.sun.wadl+xml&quot;</span>\n                <span class=\"p\">},</span>\n                <span class=\"s\">&quot;status&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;CURRENT&quot;</span><span class=\"p\">,</span>\n                <span class=\"s\">&quot;updated&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;2012-01-19T22:30:00.25Z&quot;</span>\n            <span class=\"p\">}</span>\n        <span class=\"p\">]</span>\n    <span class=\"p\">}</span>\n<span class=\"p\">}</span>\n</pre></div>\n</div>\n<p>You can obtain additional detailed information about a specific version\nby performing a <strong>GET</strong> on the base version URL. For example,\n<code class=\"docutils literal\"><span class=\"pre\">https://identity.api.rackspacecloud.com/v2.0/</span></code> is a base version URL,\nin which <code class=\"docutils literal\"><span class=\"pre\">v2.0</span></code> is the initial version of the API. All version request\nURLs end with a trailing slash (/). If the slash is omitted, the server\nmight respond with a 302 redirection request. You can add format\nextensions after the trailing slash. For example,\n<code class=\"docutils literal\"><span class=\"pre\">https://identity.api.rackspacecloud.com/v2.0/.xml</span></code> includes <code class=\"docutils literal\"><span class=\"pre\">.xml</span></code>\nas a format extension. Note that this is a special case that does not\nhold true for other API requests. In general, requests such as\n<code class=\"docutils literal\"><span class=\"pre\">/tenants.xml</span></code> and <code class=\"docutils literal\"><span class=\"pre\">/tenants/.xml</span></code> are handled equivalently.</p>\n<p><strong>Example&nbsp;3.34.&nbsp;Version Details Request</strong></p>\n<div class=\"code highlight-python\"><div class=\"highlight\"><pre>GET HTTP/1.1\nHost: identity.api.rackspacecloud.com/v1.0/\n</pre></div>\n</div>\n<p>Normal Response Code(s): 200, 203</p>\n<p>Error Response Code(s): badRequest (400), identityFault (500),\nserviceUnavailable(503)</p>\n<p>This operation does not require a request body.</p>\n<p><strong>Example&nbsp;3.35.&nbsp;Version Details Response: XML</strong></p>\n<div class=\"code highlight-python\"><div class=\"highlight\"><pre>&lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot;?&gt;\n&lt;version xmlns=&quot;http://docs.openstack.org/common/api/v1.0&quot;\n         xmlns:atom=&quot;http://www.w3.org/2005/Atom&quot;\n         id=&quot;v2.0&quot; status=&quot;CURRENT&quot; updated=&quot;2011-01-21T11:33:21-06:00&quot;&gt;\n\n     &lt;media-types&gt;\n         &lt;media-type base=&quot;application/xml&quot;\n            type=&quot;application/vnd.openstack.identity+xml;version=2.0&quot;/&gt;\n         &lt;media-type base=&quot;application/json&quot;\n            type=&quot;application/vnd.openstack.identity+json;version=2.0&quot;/&gt;\n     &lt;/media-types&gt;\n\n     &lt;atom:link rel=&quot;self&quot;\n         href=&quot;https://identity.api.rackspacecloud.com/v2.0/&quot;/&gt;\n\n    &lt;atom:link rel=&quot;describedby&quot;\n               type=&quot;application/pdf&quot;\n               href=&quot;http://docs.rackspace.com/auth/api/v2.0/auth-client-devguide-latest.pdf&quot; /&gt;\n\n    &lt;atom:link rel=&quot;describedby&quot;\n               type=&quot;application/vnd.sun.wadl+xml&quot;\n               href=&quot;http://docs.rackspacecloud.com/auth/api/v2.0/auth.wadl&quot; /&gt;\n&lt;/version&gt;\n</pre></div>\n</div>\n<p><strong>Example&nbsp;3.36.&nbsp;Version Details Response: Atom</strong></p>\n<div class=\"code highlight-python\"><div class=\"highlight\"><pre>&lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot;?&gt;\n&lt;feed xmlns=&quot;http://www.w3.org/2005/Atom&quot;&gt;\n  &lt;title type=&quot;text&quot;&gt;About This Version&lt;/title&gt;\n  &lt;updated&gt;2011-01-21T11:33:21-06:00&lt;/updated&gt;\n  &lt;id&gt;https://identity.api.rackspacecloud.com/v2.0/&lt;/id&gt;\n   &lt;author&gt;&lt;name&gt;OpenStack&lt;/name&gt;&lt;uri&gt;http://www.openstack.org/&lt;/uri&gt;&lt;/author&gt;\n   &lt;link rel=&quot;self&quot; href=&quot;http://identity.api.openstack.org/v2.0/&quot;/&gt;\n   &lt;entry&gt;\n      &lt;id&gt;http://identity.api.rackspacecloud.com/v2.0/&lt;/id&gt;\n      &lt;title type=&quot;text&quot;&gt;Version v2.0&lt;/title&gt;\n      &lt;updated&gt;2011-01-21T11:33:21-06:00&lt;/updated&gt;\n     &lt;link rel=&quot;self&quot; href=&quot;http://identity.api.rackspace.com/v2.0/&quot;/&gt;\n      &lt;link rel=&quot;describedby&quot; type=&quot;application/pdf&quot;\n           href=&quot;http://docs.rackspace.com/auth/api/v2.0/auth-client-devguide-latest.pdf&quot;/&gt;\n      &lt;link rel=&quot;describedby&quot; type=&quot;application/vnd.sun.wadl+xml&quot;\n           href=&quot;http://docs.rackspace.com/auth/api/v2.0/auth.wadl&quot;/&gt;\n      &lt;content type=&quot;text&quot;&gt;Version v2.0 CURRENT (2011-01-21T11:33:21-06:00)&lt;/content&gt;\n   &lt;/entry&gt;\n&lt;/feed&gt;\n</pre></div>\n</div>\n<p><strong>Example&nbsp;3.37.&nbsp;Version details response: JSON</strong></p>\n<div class=\"code highlight-python\"><div class=\"highlight\"><pre><span class=\"p\">{</span>\n    <span class=\"s\">&quot;version&quot;</span><span class=\"p\">:</span> <span class=\"p\">{</span>\n        <span class=\"s\">&quot;id&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;v2.0&quot;</span><span class=\"p\">,</span>\n        <span class=\"s\">&quot;links&quot;</span><span class=\"p\">:</span> <span class=\"p\">[</span>\n            <span class=\"p\">{</span>\n                <span class=\"s\">&quot;href&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;https://identity.api.rackspacecloud.com/v2.0&quot;</span><span class=\"p\">,</span>\n                <span class=\"s\">&quot;rel&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;self&quot;</span>\n            <span class=\"p\">},</span>\n            <span class=\"p\">{</span>\n                <span class=\"s\">&quot;href&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;http://docs.rackspace.com/auth/api/v2.0/auth-client-devguide-latest.pdf&quot;</span><span class=\"p\">,</span>\n                <span class=\"s\">&quot;rel&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;describedby&quot;</span><span class=\"p\">,</span>\n                <span class=\"s\">&quot;type&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;application/pdf&quot;</span>\n            <span class=\"p\">},</span>\n            <span class=\"p\">{</span>\n                <span class=\"s\">&quot;href&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;http://docs.rackspacecloud.com/auth/api/v2.0/auth.wadl&quot;</span><span class=\"p\">,</span>\n                <span class=\"s\">&quot;rel&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;describedby&quot;</span><span class=\"p\">,</span>\n                <span class=\"s\">&quot;type&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;application/vnd.sun.wadl+xml&quot;</span>\n            <span class=\"p\">}</span>\n        <span class=\"p\">],</span>\n        <span class=\"s\">&quot;media-types&quot;</span><span class=\"p\">:</span> <span class=\"p\">{</span>\n            <span class=\"s\">&quot;values&quot;</span><span class=\"p\">:</span> <span class=\"p\">[</span>\n                <span class=\"p\">{</span>\n                    <span class=\"s\">&quot;base&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;application/xml&quot;</span><span class=\"p\">,</span>\n                    <span class=\"s\">&quot;type&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;application/vnd.openstack.identity+xml;version=2.0&quot;</span>\n                <span class=\"p\">},</span>\n                <span class=\"p\">{</span>\n                    <span class=\"s\">&quot;base&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;application/json&quot;</span><span class=\"p\">,</span>\n                    <span class=\"s\">&quot;type&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;application/vnd.openstack.identity+json;version=2.0&quot;</span>\n                <span class=\"p\">}</span>\n            <span class=\"p\">]</span>\n        <span class=\"p\">},</span>\n        <span class=\"s\">&quot;status&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;CURRENT&quot;</span><span class=\"p\">,</span>\n        <span class=\"s\">&quot;updated&quot;</span><span class=\"p\">:</span> <span class=\"s\">&quot;2012-01-21T11:33:21-06:00&quot;</span>\n    <span class=\"p\">}</span>\n<span class=\"p\">}</span>\n</pre></div>\n</div>\n<p>The detailed version response contains pointers to both a human-readable\nand a machine-processable description of the API service. The\nmachine-processable description is written in the Web Application\nDescription Language (WADL).</p>\n<div class=\"admonition note\">\n<p class=\"first admonition-title\">Note</p>\n<p class=\"last\">If there is a discrepancy between the two specifications, the WADL is\nauthoritative as it contains the most accurate and up-to-date\ndescription of the API service.</p>\n</div>\n</div>\n",
    "alabaster_version": "0.7.4",
    "display_toc": false,
    "title": "Contract versions",
    "sourcename": "general-api-info/contract-versions.txt",
    "customsidebar": null,
    "current_page_name": "general-api-info/contract-versions",
    "next": {
        "link": "../extensions/",
        "title": "API Extensions"
    },
```

**Comment by ktbartholomew** [comment]
@meker12 the "Cadillac" you described is pretty much what's happening in the preparer right now. When I submit the cloud servers docs to the content service, the [resulting metadata](https://developer.rackspace.com:9000/content/https%3A%2F%2Fgithub.com%2Frackerlabs%2Fdocs-cloud-servers) looks like this:

``` json
"meta": {
    "github_edit_url": "https://github.com/rackerlabs/docs-cloud-servers/edit/master/api-docs/index.rst",
    "github_issues_url": "https://github.com/rackerlabs/docs-cloud-servers/issues",
    "collapseSections": "true"
}
```

If you follow that edit URL, it'll take you directly to the Github edit screen for the index file of the repo. 

Now, because many of our docs only appear as a single page to deconst, the only edit URL the preparer is able to present is the one for index.rst. If we were to break these docs repos into a few individual pages (something I'd eventually like us to move towards), then these edit URLs would be a lot more helpful because they would guide would-be editors a lot closer to the file that actually contains the content they want to edit.

**Comment by meker12** [comment]
That's good news.  I like Cadillacs. 

I realize for the docs using the singlehtml template, that we can really only take them to index.rst.  
In that case, I'm wondering if it would be better to just link them to the doc-root URL. The index.rst file isn't a terribly helpful place to search from.

For the user guides that provide separate topics: Will those be able to provide topic-level links to each page?

So since we have the metadata :100: ;  then we just need two template updates, right? 
-  docs-singlepage.html to add an attractive, useful **Edit on github** link to left-navigation or somewhere else easily accessible. Hopefully, linking to the Sphinx project directory at the top-level.

![singlehtml-edit-on-github](https://cloud.githubusercontent.com/assets/6391769/12629511/de2b0334-c50d-11e5-8f61-e4853140d89e.png)
-  user-guide.html to add Edit on github link to each topic?

**Comment by ktbartholomew** [comment]
@meker12 Yeah, the template changes to add the link are pretty trivial but before pulling the trigger on that, we need UX input from RED and to make sure that all the documentation repos have their `_deconst.json` files updated to look like this one: https://github.com/rackerlabs/docs-cloud-servers/blob/master/api-docs/_deconst.json

**Comment by meker12** [comment]
Got it.  We can make those _deconst.json changes now, right?

**Comment by ktbartholomew** [comment]
yep!

**Comment by meker12** [comment]
Moved description of remaining work to be completed to nexus-control issue:  See  https://github.com/rackerlabs/nexus-control/issues/295.

**Comment by ktbartholomew** [comment]
Since this issues was for adding the metadata to our content envelopes—and that's complete—I'm closing this issue. The rest of this discussion will be pretty site-specific, so should probably happen over in [rackerlabs/nexus-control](https://github.com/rackerlabs/nexus-control).

FYI @meker12

**Comment by meker12** [comment]
OK.  I moved the remaining work to this issue:    https://github.com/rackerlabs/nexus-control/issues/295
